heron/stmgr/src/cpp/manager/tmaster-client.cpp

```diff
@@ -210,31 +214,42 @@ void TMasterClient::OnHeartbeatTimer() {
 }
 
 void TMasterClient::SendRegisterRequest() {
-  CHECK(register_request_set_);
-  proto::tmaster::StMgrRegisterRequest* request =
-    new proto::tmaster::StMgrRegisterRequest(register_request_);
+  auto request = new proto::tmaster::StMgrRegisterRequest();
+
+  sp_string cwd;
+  FileUtils::getCwd(cwd);
+  proto::system::StMgr* stmgr = request->mutable_stmgr();
+  stmgr->set_id(stmgr_id_);
+  stmgr->set_host_name(stmgr_host_);
+  stmgr->set_data_port(stmgr_port_);
+  stmgr->set_local_endpoint("/unused");
+  stmgr->set_cwd(cwd);
+  stmgr->set_pid((sp_int32)ProcessUtils::getPid());
+  stmgr->set_shell_port(shell_port_);
+  for (auto iter = instances_.begin(); iter != instances_.end(); ++iter) {
+    request->add_instances()->CopyFrom(*(*iter));
+  }
+
   SendRequest(request, nullptr);
   return;
 }
 
-void TMasterClient::SetStmgrRegisterRequest(
-                                    const std::vector<proto::system::Instance*>& _instances) {
-    register_request_.Clear();
-    register_request_set_ = true;
 
-    sp_string cwd;
-    FileUtils::getCwd(cwd);
-    proto::system::StMgr* stmgr = register_request_.mutable_stmgr();
-    stmgr->set_id(stmgr_id_);
-    stmgr->set_host_name(stmgr_host_);
-    stmgr->set_data_port(data_port_);
-    stmgr->set_local_data_port(local_data_port_);
-    stmgr->set_local_endpoint("/unused");
-    stmgr->set_cwd(cwd);
-    stmgr->set_pid((sp_int32)ProcessUtils::getPid());
-    stmgr->set_shell_port(shell_port_);
+void TMasterClient::SetInstanceInfo(const std::vector<proto::system::Instance*>& _instances) {
     for (auto iter = _instances.begin(); iter != _instances.end(); ++iter) {
-      register_request_.add_instances()->CopyFrom(*(*iter));
+      auto instance = new proto::system::Instance();
```

Comment:
1. We need to Clean instances_ if it already exists
2. for (auto iter = _instances.begin(); iter != _instances.end(); ++iter) {
    auto instance = new proto::system::Instance();
    instance->CopyFrom(*iter);
    instances_.push_back(instance)
   }